providers/powerdns/diff.go

```diff
@@ -0,0 +1,145 @@
+				Msg: msgJoined,
+				F: func() error {
+					return dsp.client.Zones().RemoveRecordSetFromZone(context.Background(), dsp.ServerName, dc.Name, labelName, labelType)
+				},
+			})
+		} else {
+			// record found so create or update it
+			ttl := desiredRecords[label][0].TTL
+			var records []zones.Record
+			for _, recordContent := range desiredRecords[label] {
+				records = append(records, zones.Record{
+					Content: recordContent.GetTargetCombined(),
+				})
+			}
+			cuCorrections = append(cuCorrections, &models.Correction{
+				Msg: msgJoined,
+				F: func() error {
+					return dsp.client.Zones().AddRecordSetToZone(context.Background(), dsp.ServerName, dc.Name, zones.ResourceRecordSet{
+						Name:       labelName,
+						Type:       labelType,
+						TTL:        int(ttl),
+						Records:    records,
+						ChangeType: zones.ChangeTypeReplace,
+					})
+				},
+			})
+		}
+	}
+
+	// append corrections in the right order
+	// delete corrections must be run first to avoid correlations with existing RR
+	var corrections []*models.Correction
+	corrections = append(corrections, dCorrections...)
+	corrections = append(corrections, cuCorrections...)
+
+	return corrections, nil
+}
+
+func (dsp *powerdnsProvider) getDiff2DomainCorrections(dc *models.DomainConfig, existing models.Records) ([]*models.Correction, error) {
+	changes, err := diff2.ByRecordSet(existing, dc, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// reorder so that DELETE actions come first
+	changes = reorderChanges(changes)
+
+	var corrections []*models.Correction
+
+	for _, change := range changes {
+		labelName := canonical(change.Key.NameFQDN)
+		labelType := change.Key.Type
+
+		switch change.Type {
+		case diff2.REPORT:
+			corrections = append(corrections, &models.Correction{Msg: change.MsgsJoined})
+		case diff2.CREATE, diff2.CHANGE:
+			labelTTL := int(change.New[0].TTL)
+			records := buildRecordList(change)
+
+			corrections = append(corrections, &models.Correction{
+				Msg: change.MsgsJoined,
+				F: func() error {
+					return dsp.client.Zones().AddRecordSetToZone(context.Background(), dsp.ServerName, dc.Name, zones.ResourceRecordSet{
+						Name:       labelName,
+						Type:       labelType,
+						TTL:        labelTTL,
+						Records:    records,
+						ChangeType: zones.ChangeTypeReplace,
+					})
+				},
+			})
+		case diff2.DELETE:
+			corrections = append(corrections, &models.Correction{
+				Msg: change.MsgsJoined,
+				F: func() error {
+					return dsp.client.Zones().RemoveRecordSetFromZone(context.Background(), dsp.ServerName, dc.Name, labelName, labelType)
+				},
+			})
+		default:
+			panic(fmt.Sprintf("unhandled change.Type %s", change.Type))
+		}
+	}
+
+	return corrections, nil
+}
+
+// reorderChanges returns changes reordered so that DELETE changes run first
+func reorderChanges(changes diff2.ChangeList) diff2.ChangeList {
+	var main, pre diff2.ChangeList
+	for _, change := range changes {
+		if change.Type == diff2.DELETE {
```

Comment:
Note that .DELETE Is only used when all the records at the Label:Type are being deleted.  If, for example, a label had 3 A's and now there are only 2, the type will be .CHANGE even though (in human terms) an A record is being deleted.